ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,248 @@
+   */
+  public void handleQuery(DriverContext driverContext) {
+    long recordCreateStart = Time.monotonicNow();
+    QueryHistoryRecord record = createRecord(driverContext);
+    long recordCreateEnd = Time.monotonicNow();
+    queryHistoryQueue.add(record);
+    LOG.info("Created query history record in {}ms, current queue size: {}",
+        recordCreateEnd - recordCreateStart, queryHistoryQueue.size());
+
+    persistRecordsIfNeeded();
+  }
+
+  private void persistRecordsIfNeeded() {
+    if (needToPersist()) {
+      persistWithBatchSize(maxBatchSize);
+    } else {
+      LOG.debug("Not persisting history records yet, current queue size: {}, maxBatchSize: {}",
+          queryHistoryQueue.size(), maxBatchSize);
+    }
+  }
+
+  private void persistWithBatchSize(int batchSize) {
+    CountDownLatch latch = new CountDownLatch(1);
+    LOG.debug("Submitting a job to persist {} history records", queryHistoryQueue.size());
+    persistExecutor.submit(() -> {
+      repository.persist(queryHistoryQueue);
+      latch.countDown();
+    });
+    if (batchSize == 0) {
+      waitForAllRecordsToBePersisted(latch);
+    }
+  }
+
+  private void waitForAllRecordsToBePersisted(CountDownLatch queueEmptyCondition) {
+    try {
+      queueEmptyCondition.await();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    }
+  }
+
+  @VisibleForTesting
+  boolean needToPersist() {
+    boolean needToPersistAccordingToQueueSize = queryHistoryQueue.size() >= maxBatchSize;
+    if (needToPersistAccordingToQueueSize) {
+      LOG.info("Need to persist query history records as queue size ({}) >= batch size ({})",
+          queryHistoryQueue.size(), maxBatchSize);
+      return true;
+    }
+
+    // whether ignore the consumed memory or not
+    if (maxQueueSizeInMemoryBytes == 0) {
+      return false;
+    }
+
+    long queueMemorySize =
+        queryHistoryQueue.stream().mapToLong(QueryHistoryRecord::getEstimatedSizeInMemoryBytes).sum();
+    boolean needToPersistAccordingToQueueMemorySize = queueMemorySize > maxQueueSizeInMemoryBytes;
+    if (needToPersistAccordingToQueueMemorySize) {
+      LOG.info("Need to persist query history records as queue size in memory ({} bytes) is greater than the limit " +
+              "({} bytes)",
+          queueMemorySize, maxQueueSizeInMemoryBytes);
+    }
+    return needToPersistAccordingToQueueMemorySize;
+  }
+
+  public void stop() {
+    LOG.info("Stopping QueryHistoryService, leftover records to persist: {}", queryHistoryQueue.size());
+    persistWithBatchSize(0); // persist sync
+    persistExecutor.shutdownNow();
+  }
+
+  private QueryHistoryRecord createRecord(DriverContext driverContext) {
+    return new QueryHistoryRecordEnricher(driverContext, serviceContext, SessionState.get(),
+        SessionState.getPerfLogger()).createRecord();
+  }
+
+  @VisibleForTesting
+  public QueryHistoryRepository getRepository(){
```

Comment:
Intellij reports it as unused

Reply:
double-check, used in TestIcebergRepository
```
    IcebergRepository repository = (IcebergRepository) service.getRepository();
```